tidy3d/components/geometry/base.py

```diff
@@ -2011,6 +2161,344 @@ def _surface_area(self, bounds: Bound) -> float:
+        with the same shape which is ``True`` for every point in zip(x, y, z) that is inside the
+        volume of the :class:`Geometry`, and ``False`` otherwise.
+
+        Parameters
+        ----------
+        x : np.ndarray[float]
+            Array of point positions in x direction.
+        y : np.ndarray[float]
+            Array of point positions in y direction.
+        z : np.ndarray[float]
+            Array of point positions in z direction.
+
+        Returns
+        -------
+        np.ndarray[bool]
+            ``True`` for every point that is inside the geometry.
+        """
+        x = np.array(x)
+        y = np.array(y)
+        z = np.array(z)
+        xyz = np.dot(self.inverse, np.vstack((x.flat, y.flat, z.flat, np.ones(x.size))))
+        if xyz.shape[1] == 1:
+            # TODO: This "fix" is required because of a bug in PolySlab.inside (with non-zero sidewall angle)
+            return self.geometry.inside(xyz[0][0], xyz[1][0], xyz[2][0]).reshape(x.shape)
+        return self.geometry.inside(xyz[0], xyz[1], xyz[2]).reshape(x.shape)
+
+    def _volume(self, bounds: Bound) -> float:
+        """Returns object's volume within given bounds."""
+        # NOTE (Lucas): Bounds are overestimated.
+        vertices = np.dot(self.inverse, self._vertices_from_bounds(bounds))[:3]
+        inverse_bounds = (tuple(vertices.min(axis=1)), tuple(vertices.max(axis=1)))
+        return abs(np.linalg.det(self.transform)) * self.geometry.volume(inverse_bounds)
+
+    def _surface_area(self, bounds: Bound) -> float:
+        """Returns object's surface area within given bounds."""
+        log.warning("Surface area of transformed elements cannot be calculated.")
+        return 0.0
+
+    @staticmethod
+    def translation(x: float, y: float, z: float) -> MatrixReal4x4:
+        """Return a translation matrix.
+
+        Parameters
+        ----------
+        x : float
+            Translation along x.
+        y : float
+            Translation along y.
+        z : float
+            Translation along z.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        return np.array(
+            [
+                (1.0, 0.0, 0.0, x),
+                (0.0, 1.0, 0.0, y),
+                (0.0, 0.0, 1.0, z),
+                (0.0, 0.0, 0.0, 1.0),
+            ],
+            dtype=float,
+        )
+
+    @staticmethod
+    def scaling(x: float, y: float = None, z: float = None) -> MatrixReal4x4:
+        """Return a scaling matrix.
+
+        Parameters
+        ----------
+        x : float
+            Scaling factor along x.
+        y : float = None
+            Scaling factor along y. If ``None``, use the value for x.
+        z : float = None
+            Scaling factor along z. If ``None``, use the value for x.
+
+        Returns
+        -------
+        numpy.ndarray
+            Transform matrix with shape (4, 4).
+        """
+        if y is None:
```

Comment:
what worries me about this is that people might try
```
scaling(x=2)
```
and not realize that this is automatically setting y and z.  
```
scaling(y=2)
```
on the other hand will fail.

It's probably ok, but there's a bit of asymmetry here that I am not sure about, maybe if it were me I'd say all to be optional and = 1.0 by default? what is the reasoning here? is scaling all x, y, z far more common? and expecting the user to do it via positional argument rather than keyword argument?

Reply:
Correct. I believe scaling is usually applied homogeneously in all dimensions, but I see the problem. I think having all arguments default to 1 makes more sense.